File: frontend/src/models/vehicle/SemiTrailer.ts
export interface SemiTrailerSummary {
    id: number

    registration: string

    brand: string | null
    model: string | null
}

export interface SemiTrailerDetails extends SemiTrailerSummary {
    trailerType: string | null

    maxSpeed: number | null

    emptyWeightKg: number | null
    grossVehicleWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null

    externalId: string | null
    externalSource: string

    createdAt: string
    updatedAt: string
}

export interface CreateSemiTrailerRequest {
    externalId: string | null
    externalSource: string

    registration: string

    brand: string | null
    model: string | null

    trailerType: string | null

    maxSpeed: number | null

    emptyWeightKg: number | null
    grossVehicleWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null
}

export interface UpdateSemiTrailerRequest {
    registration: string

    brand: string | null
    model: string | null

    trailerType: string | null

    maxSpeed: number | null

    emptyWeightKg: number | null
    grossVehicleWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null

    depreciationStartDate: string | null
    depreciationEndDate: string | null
}

export interface SemiTrailerFormData {
    registration: string
    brand: string | null
    model: string | null

    maxSpeed: number | null

    trailerType: string | null

    emptyWeightKg: number | null
    grossVehicleWeightKg: number | null

    heightCm: number | null
    widthCm: number | null
    lengthCm: number | null

    axleCount: number | null

    purchaseCost: number | null
    depreciationStartDate: string | null
    depreciationEndDate: string | null
}

File: backend/src/main/java/com/optiroute/backend/entity/Vehicle/SemiTrailer.java
package com.optiroute.backend.entity.vehicle;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;

import com.optiroute.backend.entity.EntityUtils;

@Entity
@Table(name = "semi_trailer")
@Getter
@Setter
public class SemiTrailer extends EntityUtils {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "external_id")
    private String externalId;

    @Column(name = "external_source")
    private String externalSource = "MANUAL";

    @Column(nullable = false, unique = true, length = 15)
    private String registration;

    @Column(length = 100)
    private String brand;

    @Column(length = 100)
    private String model;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "trailer_type_id")
    private TrailerType trailerType;

    private Integer maxSpeed;

    private Integer emptyWeightKg;

    private Integer grossVehicleWeightKg;

    private Integer heightCm;

    private Integer widthCm;

    private Integer lengthCm;

    private Integer axleCount;

    @Column(precision = 12, scale = 2)
    private BigDecimal purchaseCost;

    private LocalDate depreciationStartDate;

    private LocalDate depreciationEndDate;
}